pkg/providers/property/data_akamai_search.go

```diff
@@ -0,0 +1,115 @@
+package property
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/akamai/AkamaiOPEN-edgegrid-golang/v2/pkg/papi"
+	"github.com/akamai/terraform-provider-akamai/v2/pkg/akamai"
+	"github.com/akamai/terraform-provider-akamai/v2/pkg/tools"
+)
+
+func dataSourceSearch() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceSearchRead,
+		Schema: map[string]*schema.Schema{
+			"key": {
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "Key must have one of three values: \"edgeHostname\", \"hostname\" or \"propertyName\"",
+				ValidateDiagFunc: tools.IsNotBlank,
+			},
+			"value": {
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "Value of Search",
+				ValidateDiagFunc: tools.IsNotBlank,
+			},
+			"properties": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "List of properties",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"account_id":        {Type: schema.TypeString, Computed: true},
+						"asset_id":          {Type: schema.TypeString, Computed: true},
+						"contract_id":       {Type: schema.TypeString, Computed: true},
+						"group_id":          {Type: schema.TypeString, Computed: true},
+						"property_id":       {Type: schema.TypeString, Computed: true},
+						"property_version":  {Type: schema.TypeInt, Computed: true},
+						"property_name":     {Type: schema.TypeString, Computed: true},
+						"edge_hostname":     {Type: schema.TypeString, Computed: true},
+						"hostname":          {Type: schema.TypeString, Computed: true},
+						"production_status": {Type: schema.TypeString, Computed: true},
+						"staging_status":    {Type: schema.TypeString, Computed: true},
+						"updated_by_user":   {Type: schema.TypeString, Computed: true},
+						"updated_date":      {Type: schema.TypeString, Computed: true},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceSearchRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
```

Comment:
Hi @Slonimskaia,
I actually think that an empty result is better than an error.
For example: 
Let's say you want to know how many properties are leveraging few edge hostnames, and then operate some other changes based on this piece of information.
you'll write something like:

```hcl
data "akamai_property_search" "test" {
  for_each = var.hosts
  key   = "edgeHostname"
  value = each.key
}
// some other code that leverage on akamai_property_search.test.*.properties
```
If we throw an error in case of not found, then the code could break even when it could be expected that some edge hostnames aren't listed in any properties yet.

Same problem will be there even if you use this data source in the same code of a brand new property.
Can we please not throw an error?

Reply:
Hi @Fuochi ,

I initially thought the same as you, but then I remembered that this is the approach we follow in our data sources (please check in the `property` subprovider, which is the one we always take as reference: `data_akamai_contracts.go`, `data_akamai_cp_code.go`, `data_akamai_property.go`, etc.).

Regards,

Roberto

Reply:
Hi @robertolopezlopez,
Thank you for giving me some context. Still, I think that since this is the implementation of search, it should return an empty value. Isn't `data_akamai_properties.go` doing the same?

Reply:
Hi @Fuochi,

The problem here is that if user try to run _terraform apply_ on key _propertyName_ with not existing property, then state file is created and id is assigned. 
![Screen Shot 2022-03-04 at 11 57 14 AM](https://user-images.githubusercontent.com/43030708/156751022-b6f4cc7f-2d18-47f9-b7e2-b43d8844916a.jpg)
It is not expected behavior.
Let me verify it for edge hostname and hostname. I will test your implementation and come back with suggestion.

Reply:
@Fuochi, yes, data_akamai_properties.go returns a Not found error, if property doesn't exist

Reply:
@Fuochi I put into account only `data_akamai_property.go` data source. I need some time to verify it with `data_akamai_properties.go`. Maybe changes which I suggested above are not necessary.

Reply:
@Slonimskaia no problem!
Thank you for the time you and @robertolopezlopez are spending in this.

Reply:
I discussed this topic with my colleagues. We have a couple of suggestions, which should be addressed. Since only one property can be associated with edge hostname and the same is for hostname, at this moment the result  of this data source is list of property versions.The same is for property name, if you search by name, you get versions only for one property. I suggest to return not list of versions, but only the latest one.

TODO:
1. In data source return only the latest property version
2. Field "properties" in schema should be replaced with property fields(account_id, asset_id and etc.). Property fields should be at the same level as _key_ and _value_ fields:
![Screen Shot 2022-03-07 at 2 30 19 PM](https://user-images.githubusercontent.com/43030708/157043838-48a546c0-4d42-45ab-bacc-3b4162bc34a8.jpg)
3. If client.SearchProperties(ctx, request) in read function returns an empty result, then return ErrPropertyNotFound. The same approach is used in data_akamai_property.go:
![Screen Shot 2022-03-07 at 2 43 19 PM](https://user-images.githubusercontent.com/43030708/157045821-78b90f47-2919-45fc-9410-ce7a9ee15e2d.jpg)
![Screen Shot 2022-03-07 at 2 39 39 PM](https://user-images.githubusercontent.com/43030708/157045840-a126d6f4-ad5c-46d1-a9f1-a645425b51e0.jpg)
For data_akamai_properties.go it works differently, since in this resource we search properties by group id and contract id.

5. Please rename files and everything what is related from _properties_ to _property_ . Sorry for this request again, but since implementation should be changed, then names should be changed also. 
6. Change tests according to new implementation

Reply:
Hi @Slonimskaia,
Unfortunately, I have to disagree with you. I don't think that the first assumption is entirely true.
Even though there is only one property per property name, surely there could be multiple properties using the same edge hostname. Also, in some edge cases, it's possible that multiple properties can have the same hostname as well, the only bound is that only one of them can be active in a given network (i.e. staging and production).

Right now, it returns a list of property versions, for each property that matches the criteria, it can provide up to 3 versions: last, staging active, and production active. It can be useful to have all of them in this list.

If you prefer, we can change the name of the data source to something like `akamai_property_versions_search` and the `properties` field to `property_versions` or something like that.

What do you think?

Reply:
@Fuochi you are right, that hostname and edgeHostname can be associated with multiple properties, but only one associated property can be active. Now I think, that what I suggested above, doesn't make sense, if expected result from this data source is all the versions of properties(active and inactive).  Please ignore my previous suggestions.
I would suggest only next changes:
1. I think, it make sense to return ErrPropertyNotFound, If client.SearchProperties(ctx, request) in read function returns an empty result. What do you think? 
2. I suppose, that ```akamai_properties_search``` name can stay as it is now, instead of ```akamai_property_versions_search```, since this data source returns not only versions for property, but versions for different properties(active and inactive). Please rename, if you think it is needed. Do not forget add changes in properties_search.md. I see two inconsistencies there for now, in lines 20 and 26:
![Screen Shot 2022-03-08 at 12 32 42 PM](https://user-images.githubusercontent.com/43030708/157230563-123658a7-a47d-442f-9c6b-80fde86b71b6.jpg)

Reply:
Hi @Slonimskaia,
Thank you for understanding my point. About the two suggestions posted here:

1. I still don't like the idea of throwing an error instead of saving an empty list in the state. Let me explain why I made this PR. We'd like to use this data source to check if a specific hostname is already used by any property and take actions on other properties only if it's not used yet. We'll have something like:
 `count = length(data.akamai_properties_search.property) == 0 ? 1 : 0`
in different resources. This could be easily checked if the data source has an empty list. It's not that easy to do the same with an error.
However, if you put this as a requirement to have this PR approved, I can implement the suggested change, then we'll try to find a workaround.

3. Sure, just fixed. Thanks

Reply:
@Fuochi I see your point of view about first suggestion. Thank you for the explanation. Let me retest this data source and we can merge it